Source/WebDriver/WebDriverService.cpp

```diff
@@ -2458,4 +2606,35 @@ void WebDriverService::takeElementScreenshot(RefPtr<JSON::Object>&& parameters,
     });
 }
 
+#if ENABLE(WEBDRIVER_BIDI)
+
+void WebDriverService::clientDisconnected(WebSocketMessageHandler::Connection connection)
+{
+    // https://w3c.github.io/webdriver-bidi/#handle-a-connection-closing
+    // To handle a connection closing given a WebSocket connection connection:
+    // 1. If there is a BiDi session associated with connection connection:
+    //   1.1 Let session be the BiDi session associated with connection connection.
+    if (m_bidiServer.getSessionForConnection(connection)) {
+        //   1.2. Remove connection from session’s session WebSocket connections.
+        m_bidiServer.removeConnectionAssociatedToSession(connection);
+    } else if (m_bidiServer.isConnectionNotAssociatedToSession(connection)) {
+        // 2. Otherwise, if WebSocket connections not associated with a session contains connection, remove connection from that set.
+        m_bidiServer.removeConnectionNotAssociatedWithSession(connection);
+    }
+
+    // Note from spec: This does not end any session.
+}
+
+void WebDriverService::onBrowserKilled(const String& sessionID)
+{
+    if (m_session && m_session->id() == sessionID) {
+        // Disconnect the WebSocket connection given session
+        auto connection = m_bidiServer.getConnectionForSession(sessionID);
+        m_bidiServer.disconnectSession(sessionID);
+        clientDisconnected((*connection).get());
```

Comment:
Is this safe to deref without a null check?

Reply:
Added check.